### ISSUE TITLE
Support for the ppc64le architecture via existing patches

### ISSUE THREAD

**Issue by shawnanastasio** [electron/electron#15691]
**Is your feature request related to a problem? Please describe.**
Currently Electron does not support the ppc64le architecture and therefore can't be used on machines like the libre Talos II workstation from Raptor Computing Systems.

On this front, myself and others have recently ported Chromium to ppc64le. We are currently attempting to get them upstream in Chromium, but unfortunately Google has been slow in acknowledging them.

In the meantime, I have integrated the patches into Electron's patch hook which allows the project to be built and run on ppc64le systems[1]. Incorporating these changes would allow users on the ppc64le platform to benefit from Electron's broad ecosystem without needing to wait for Google.

**Describe the solution you'd like**
While the Chromium patches pend upstream, I would like for them to be incorporated into Electron's build system. I have already done this in a fork[1], and with the Electron team's guidance, I would like for them to be merged so that official Electron binaries can be provided and the existing ecosystem tooling can take advantage of them.

**Describe alternatives you've considered**
Ideally, the patches will be accepted into upstream Chromium and Electron will have automatic support as a downstream. As stated above, though, there are some delays on that front and I believe the ppc64 community would greatly benefit from Electron support in the meantime.

**Additional context**
The Chromium patchset is actively maintained by me to track upstream, so the Electron team would simply need to pull my changes on Chromium version bumps. Ideally, this would introduce as little maintenance overhead as possible.

I have also written build instructions for ppc64le machines[2].

On the unit test front, there is an issue when running the `npm i --nodedir=../../out/Debug/gen/node_headers` step as described in the Electron documentation. The issue does not appear to be directly related to electron, but rather an to npm dependency of the unit tests. I have attached the output of this command in hopes that someone more familiar with the ecosystem can provide some insight[3].

I believe introducing ppc64le support in Electron would greatly benefit the users and developers of the ppc64 community, as well as the libre software community as a whole.

[1] https://github.com/shawnanastasio/electron
[2] https://wiki.raptorcs.com/wiki/Porting/Chromium/Electron
[3] https://paste.fedoraproject.org/paste/cYjtZcqxlcV5RwLPsjAxWQ

**Comment by madscientist159** [comment]
I can confirm we can provide free CI and build machine access via IntegriCloud to select libre software projects.  This one would definitely qualify; if you're interested, contact sales@integricloud.com and point them to this issue.

**Comment by leo-lb** [comment]
I have created another ppc64le fork using a different strategy at https://github.com/leo-lb/electron
I have created forks for every dependency that needed modifications and integrated these dependencies within `DEPS` files for use with gclient. These changes are mostly transparent to Electron, almost all the changes are in Chromium.

**Comment by darkbasic** [comment]
Thanks for your effort, this is really appreciated. Hopefully they will decide to merge those patches before Chrome does.

**Comment by miniak** [comment]
There is no progress here, does it make sense to keep the issue open?

**Comment by darkbasic** [comment]
Isn't the point of a gh issue to actually **track** issues? Closing them doesn't make the issue magically go away...

**Comment by madscientist159** [comment]
Agreed, and the message closing the sends is "electron apps require proprietary firmware to run".  Is that really a great idea?

**Comment by miniak** [comment]
ok :) you're right, let's keep it then

**Comment by darkbasic** [comment]
Does anybody still maintain the electron ppc64le patchset? I've finally got my Raptor system repaired and I absolutely need to get at least vscode running to keep using it as my main workstation, but unfortunately it looks like the patching effort has slowed down.
I remember using leo-lb's branch (https://github.com/leo-lb/vscode) but unfortunately that doesn't look active as well.

**Comment by darkbasic** [comment]
If someone is interested I'm now maintaining the ppc64le port of the Electron and Slack Gentoo ebuilds: https://github.com/PF4Public/gentoo-overlay/pull/134

**Comment by lex-ibm** [comment]
Hi!
I'm interested in that set of patches going to upstream chromium. The issue at [1242633](https://chromium-review.googlesource.com/c/crashpad/crashpad/+/1242633) seems to have gone quiet since 2021. Is that still the main effort to get the patches upstream?
Also @darkbasic, thanks for the continued effort!

**Comment by tpearson-ssc** [comment]
A full set of working, field tested patches are maintained in the Debian project here:

https://salsa.debian.org/chromium-team/chromium/-/tree/master/debian/patches/ppc64le

Google has been most unpleasant regarding upstreaming, eventually dictating that due to the outcome of a secret internal meeting they won't be accepting patches.  As a result, I'm working on a way to maintain this patchset somewhere where more than just Debian can easily benefit from it, including CI/CD going forward, but as this is a large task the various ppc64el maintainers should probably try to collaborate on that task.  Solid has resources available to assist with this, including native POWER runners on its public Gitlab instance.

**Comment by runlevel5** [comment]
> As a result of Google's decision and renewed interest in multi-distro support, I'm working on a way to maintain this patchset somewhere where more than just Debian can easily benefit from it, including the necessary CI/CD going forward. Since this is a large task, I would strongly recommend that the various ppc64el maintainers try to collaborate, and I'm happy to lead that effort. Solid also has resources available to assist with this, including native POWER runners on its public Gitlab instance, and we are also working on v8 feature parity with amd64/arm64. Thankfully the v8 maintainers have been willing to work with us and accept our patches, it's just Google itself for the Chromium browser that has an undisclosed block on merging patches.

I am too interested in maintaining a package for Fedora. Please share with me more details once the patchset has a home.

**Comment by lex-ibm** [comment]
@runlevel5 great news! Looks like Fedora does have some patches for PPC64LE [chromium](https://src.fedoraproject.org/rpms/chromium). I haven't looked too much into it, but they might be the same as the ones in Debian and Raptor Engineering.

@tpearson-ssc thanks for all the hard work! I've been able to build electron thanks to those patches! I don't think I'm the right person to talk to Google directly, but I'll share this with some folks I know might be interested.

**Comment by runlevel5** [comment]
> Was there any interest in a bimonthly sync call for the various maintainers of Chrome/v8 on ppc64el?

I am interested

> Also, does it make sense to have a merged tree automatically generated so that edits can be made directly in Git vs. only as patches?

Sounds like a good idea to me

**Comment by sharkcz** [comment]
I like the multi-distro idea, big +1

**Comment by darkbasic** [comment]
If we're going to make a distro agnostic patchset I would like to point out that Gentoo has a couple of additional patches on top of Timothy's patchset: https://github.com/gentoo/gentoo/blob/master/www-client/chromium/chromium-126.0.6478.114.ebuild#L85C27-L85C58

Not sure why they're not needed on Debian but it's mostly random compilation issues that I ended up piling on top of the debian patchset.

**Comment by tpearson-ssc** [comment]
Perfect!  I'm setting up the patch-applied tree here, just waiting for data to transfer, could be a day or so still:

https://gitlab.solidsilicon.io/public-development/open-source/chromium/chromium-openpower

For the sync call, I'd propose first and third week of the month to align somewhat with Google's release schedule.  We can take that scheduling offline, Email me directly at tpearson@solidsilicon.com and we'll figure something out.

@darkbasic If those are ppc64el specific, can you create a merge request on https://gitlab.solidsilicon.io/public-development/open-source/chromium/openpower-patches to add them in?  Debian turns some things on / off so it's entirely possible the code paths simply aren't hit.

**Comment by lex-ibm** [comment]
I've been building electron for PowerPC64 Little Endian for a while now without issues.
I'm going to leave here the build script I'm using just in case someone needs it:
- [lex-ibm/electron-ppc64le-build-scripts](https://github.com/lex-ibm/electron-ppc64le-build-scripts)
I might work on getting some workflow to release binaries, but that's a stretch.

**Comment by runlevel5** [comment]
It's been few years since, I know it might be a bit futile to ask for official support from the team but I guess we could try again. It would mean a world to the PowerPC community if the patchset is merged into the main trunk

**Comment by miniak** [comment]
> It's been few years since, I know it might be a bit futile to ask for official support from the team but I guess we could try again. It would mean a world to the PowerPC community if the patchset is merged into the main trunk

are you able to create a PR for this? I guess adding it as an unofficial platform could be doable. cc @electron/wg-releases, @electron/wg-upgrades

**Comment by runlevel5** [comment]
> Are you able to create a PR with the changes rebased for the current main?

@miniak sure I can! I will get onto that this weekend

**Comment by runlevel5** [comment]
@lex-ibm i am thinking to adapt your https://github.com/lex-ibm/electron-ppc64le-build-scripts but I guess if you have capacity right now, could you please help me prepare a Pull Request with all patches required?

**Comment by darkbasic** [comment]
@miniak if there really is some upstream interest I'm pretty sure @madscientist159 would be more than happy to open such PR himself.

**Comment by lex-ibm** [comment]
Woho!
@runlevel5 sure! I'll give it a look over the weekend

**Comment by JeremyRand** [comment]
@lex-ibm Status?

**Comment by lex-ibm** [comment]
@JeremyRand I have been wanting to come back to this but unfortunately I've had a lot of work on my plate. Also, I remember last build I did of electron had some issues around ffmpeg that I didn't get to look into. 

So, yeah, I'm not currently working on this and I can't promise I'll look into it any time soon...

Last build I did was [v37.2.3](https://github.com/lex-ibm/electron-ppc64le-build-scripts/releases/tag/v37.2.3) ([this commit](https://github.com/lex-ibm/electron-ppc64le-build-scripts/commit/d66b5002ab4c3bfcfd26f792e80ce884105428f2))

**Comment by JeremyRand** [comment]
OK thanks @lex-ibm. @runlevel5 do you want to work on a PR, or should we try to get Timothy to handle it?

**Comment by runlevel5** [comment]
@JeremyRand I am more than happy to assist but it would be better if Timothy or someone who have familiarized with the codebase to have a go

**Comment by JeremyRand** [comment]
@tpearson-ssc is this something that you have the time to work on currently?

**Comment by darkbasic** [comment]
Electron already works with Timothy patches. The main problem is that Google refuses to merge its patches in Chromium because they don't care about a supposedly niche architecture.

**Comment by runlevel5** [comment]
> Electron already works with Timothy patches. The main problem is that Google refuses to merge its patches in Chromium because they don't care about a supposedly niche architecture.

It's okay. We will get there eventually. So far I have seen that introducing support downstream has worked very well, for example Fedora now ships ppc64le Chromium with Timothy's patches. The more upstream support we could get, the easier it will be to get other upstream projects on board.

**Comment by lex-ibm** [comment]
Something like this?
https://github.com/lex-ibm/electron/tree/feat/ppc64-support
That is just my playground for the latest build I pushed, just playing around with the idea of pushing the patches.

**Comment by lex-ibm** [comment]
I guess the idea would be to cross-compile from x86-64 just as arm64 is... am I wrong?

**Comment by runlevel5** [comment]
> I guess the idea would be to cross-compile from x86-64 just as arm64 is... am I wrong?

Yeah ideally. If cross-compilation is not an option, we could always build it natively on real hardware. See https://github.com/IBM/actionspz

**Comment by lex-ibm** [comment]
Well... it does cross-compile, but fails with:
```
#
# Fatal error in ../../v8/src/heap/heap.cc, line 3331
# Debug check failed: roots.is_initialized(RootIndex::kFreeSpaceMap) implies IsMap(HeapObject::FromAddress(free_space.Address()) ->map(heap->isolate())).
#
#
#
#FailureMessage Object: 0x7fffc8ba3a90
```
